File: servers/Services/AcasService/Models/ErrorGroup.cs
using Amazon.DynamoDBv2.DataModel;
using System.ComponentModel.DataAnnotations;
using AcasService.Dev;

namespace AcasService.Models;

[DynamoDBEntity("ErrorGroupTableName")]
public class ErrorGroup
{
    [Key]
    public string Id { get; set; } = string.Empty;

    public string ProblemId { get; set; } = string.Empty;

    public string ExamId { get; set; } = string.Empty;

    public string ErrorSignature { get; set; } = string.Empty;

    public JPlagStatus JPlagStatus { get; set; } = JPlagStatus.PENDING; 

    public List<string> SubmissionIds { get; set; } = new List<string>();

    public List<JPlagMatch> JPlagResults { get; set; } = new List<JPlagMatch>();
    
    public DateTime CreatedDate { get; set; } = DateTime.UtcNow;
}

public class JPlagMatch
{
    public string Submission1Id { get; set; } = string.Empty;
    public string Submission2Id { get; set; } = string.Empty;
    public float SimilarityScore { get; set; } 
    public List<JPlagMatchDetail> Details { get; set; } = new List<JPlagMatchDetail>();
}

public class JPlagMatchDetail
{
    public int StartLine1 { get; set; }
    public int EndLine1 { get; set; }
    public int StartLine2 { get; set; }
    public int EndLine2 { get; set; }
    public int Tokens { get; set; }
}

public enum JPlagStatus
{
    PENDING,
    RUNNING,
    COMPLETED,
    FAILED
}

File: web_app/types/error-group.ts
export interface ErrorGroupSummary {
  id: string;
  problemId: string;
  examId: string;
  errorSignature: string;
  jPlagStatus: 'PENDING' | 'RUNNING' | 'COMPLETED' | 'FAILED';
  submissionIds: string[];
  jPlagResults: JPlagMatchSummary[];
  createdDate: string;
}

export interface JPlagMatchSummary {
  submission1Id: string;
  submission2Id: string;
  similarityScore: number;
}

export interface ErrorGroupDetail extends ErrorGroupSummary {
  jPlagResultsDetailed: JPlagMatchDetailGroup[];
}

export interface JPlagMatchDetailGroup extends JPlagMatchSummary {
  details: MatchLineDetail[];
}

export interface MatchLineDetail {
  startLine1: number;
  endLine1: number;
  startLine2: number;
  endLine2: number;
  tokens: number;
}

export interface ErrorGroupRequest {
  examId: string;
  problemId: string;
  groupIds?: string[];
  minTokenMatch?: number;
  minSimilarity?: number;
  excludeBaseCode?: boolean;
}
